flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/SchedulingUtils.java

```diff
@@ -0,0 +1,219 @@
+						throw new CompletionException(
+							new FlinkException(
+								String.format("Could not deploy execution %s.", execution),
+								t));
+					}
+				}
+			})
+			// Generate a more specific failure message for the eager scheduling
+			.exceptionally(
+				(Throwable throwable) -> {
+					final Throwable strippedThrowable = ExceptionUtils.stripCompletionException(throwable);
+					final Throwable resultThrowable;
+					if (strippedThrowable instanceof TimeoutException) {
+						int numTotal = allAllocationsFuture.getNumFuturesTotal();
+						int numComplete = allAllocationsFuture.getNumFuturesCompleted();
+
+						String message = "Could not allocate all requires slots within timeout of "
+							+ executionGraph.getAllocationTimeout() + ". Slots required: "
+							+ numTotal + ", slots allocated: " + numComplete
+							+ ", previous allocation IDs: " + allPreviousAllocationIds;
+
+						StringBuilder executionMessageBuilder = new StringBuilder();
+
+						for (int i = 0; i < allAllocationFutures.size(); i++) {
+							CompletableFuture<Execution> executionFuture = allAllocationFutures.get(i);
+
+							try {
+								Execution execution = executionFuture.getNow(null);
+								if (execution != null) {
+									executionMessageBuilder.append("completed: " + execution);
+								} else {
+									executionMessageBuilder.append("incomplete: " + executionFuture);
+								}
+							} catch (CompletionException completionException) {
+								executionMessageBuilder.append("completed exceptionally: "
+									+ completionException + "/" + executionFuture);
+							}
+
+							if (i < allAllocationFutures.size() - 1) {
+								executionMessageBuilder.append(", ");
+							}
+						}
+
+						message += ", execution status: " + executionMessageBuilder.toString();
+
+						resultThrowable = new NoResourceAvailableException(message);
+					} else {
+						resultThrowable = strippedThrowable;
+					}
+
+					throw new CompletionException(resultThrowable);
+				});
+	}
+
+	/**
+	 * Returns the result of {@link #computePriorAllocationIds(Collection)},
+	 * but only if the scheduling really requires it.
+	 * Otherwise this method simply returns an empty set.
+	 */
+	static private Set<AllocationID> computePriorAllocationIdsIfRequiredByScheduling(
+		final Collection<ExecutionVertex> vertices,
+		final SlotProvider slotProvider) {
+		// This is a temporary optimization to avoid computing all previous allocations if not required
+		// This can go away when we progress with the implementation of the Scheduler.
+		if (slotProvider instanceof Scheduler &&
+			((Scheduler) slotProvider).requiresPreviousExecutionGraphAllocations()) {
+
+			return computePriorAllocationIds(vertices);
+		} else {
+			return Collections.emptySet();
+		}
+	}
+
+	/**
+	 * Computes and returns a set with the prior allocation ids for given execution vertices.
+	 */
+	static private Set<AllocationID> computePriorAllocationIds(final Collection<ExecutionVertex> vertices) {
```

Comment:
![image](https://user-images.githubusercontent.com/1681921/60166183-f5a27f00-9800-11e9-8e0c-ac89038d72fa.png)

Reply:
Will fix it.